src/limitedmap.h

```diff
@@ -26,68 +26,82 @@ class limitedmap
+        for (; it != it_last; ++it) {
+            if (it->second == map_it) {
+                return it;
+            }
+        }
+        return rmap.end();
+    }
+
 public:
-    limitedmap(size_type nMaxSizeIn = 0) { nMaxSize = nMaxSizeIn; }
+    LimitedMap(size_type nMaxSizeIn) {
+        assert(nMaxSizeIn > 0);
+        nMaxSize = nMaxSizeIn;
+    }
     const_iterator begin() const { return map.begin(); }
     const_iterator end() const { return map.end(); }
     size_type size() const { return map.size(); }
     bool empty() const { return map.empty(); }
     const_iterator find(const key_type& k) const { return map.find(k); }
     size_type count(const key_type& k) const { return map.count(k); }
-    void insert(const value_type& x)
-    {
-        std::pair<iterator, bool> ret = map.insert(x);
-        if (ret.second) {
-            if (nMaxSize && map.size() == nMaxSize) {
-                map.erase(rmap.begin()->second);
-                rmap.erase(rmap.begin());
+
+    bool insert(const value_type& x) {
+        if (map.find(x.first) != map.end()) {
+            return false;
+        }
+
+        if (map.size() == nMaxSize) {
+            const V& cur_min = rmap.begin()->first;
+            if (x.second <  cur_min) {
+                return false;
             }
-            rmap.insert(make_pair(x.second, ret.first));
+            map.erase(rmap.begin()->second);
+            rmap.erase(rmap.begin());
         }
-        return;
+
+        std::pair<iterator, bool> ret = map.insert(x);
+        assert(ret.second);
+        // in zend, we expect to insert increasing timestamps most of the times,
+        // so we suggest std::multimap to insert at the end.
+        rmap.insert(rmap.end(), make_pair(x.second, ret.first));
+        return true;
     }
-    size_t erase(const key_type& k)
-    {
+
+    size_t erase(const key_type& k) {
         iterator itTarget = map.find(k);
         if (itTarget == map.end())
             return 0;
-        std::pair<rmap_iterator, rmap_iterator> itPair = rmap.equal_range(itTarget->second);
-        for (rmap_iterator it = itPair.first; it != itPair.second; ++it) {
-            if (it->second == itTarget) {
-                rmap.erase(it);
-                map.erase(itTarget);
-                return 1; // be consistent with std::map return value
-            }
-        }
-        // Shouldn't ever get here
-        assert(0);
-        return 0;
+        rmap_iterator rit = find_rmap_key_value_pair(itTarget);
+        assert(rit != rmap.end());
+        rmap.erase(rit);
+        map.erase(itTarget);
+        return 1; // be consistent with std::map return value
     }
-    void update(const_iterator itIn, const mapped_type& v)
-    {
-        // TODO: When we switch to C++11, use map.erase(itIn, itIn) to get the non-const iterator.
-        iterator itTarget = map.find(itIn->first);
+
+    void update(const_iterator itIn, const mapped_type& v) {
```

Comment:
Possibly `update` method can have a `bool` type return value (similarly to `insert` method) in order to inform if an update has been actually performed or not; btw, this return value wouldn't be considered in current implementation, so feel free to ignore this comment.

Reply:
Yeah makes sense, updated accordingly, thanks.